tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -688,8 +695,15 @@ const DeviceCommand EnqueueWriteShardedBufferCommand::create_buffer_transfer_ins
 
     uint32_t num_cores = this->buffer.num_cores();
     uint32_t shard_size = this->buffer.shard_spec().size();
-    //TODO: for now all shards are same size of pages
-    vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
+    //TODO: for now all shards are same size of pages, except last core
+    vector<uint32_t> num_pages_in_shards(num_cores -1, shard_size);
```

Comment:
num_cores - 1 instead of num_cores -1

Reply:
we are tying to get rid of push_back code and reserve the amount of space you need AOT. Cleaner to do 

```c++
vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
if (edge_case_logic) {
    num_pages_in_shards[num_pages_in_shards.size() - 1] = edge_case_value;
}
```

Also, for readability, shard_size units are ambiguous. Typically I think of size in bytes in dispatch land, but it seems here that it means num pages in a shard (maybe make that more explicit).

Also, modulo is expensive. If you know num_pages() <= num_cores * shard_size, a cheaper way to compute the edge case is 

```c++
this->buffer.num_pages() - (num_cores - 1) * shard_size;
```

So all in all, the code would look like this

```c++
vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
uint32_t num_pages_in_shards_upper_bound = num_cores * shard_num_pages;
if (this->buffer.num_pages() != num_pages_in_shards_upper_bound) {
   num_pages_in_shards[num_pages_in_shards.size() - 1] = this->buffer.num_pages() - (num_pages_in_shard_upper_bound - shard_size);
}
```